### PR TITLE
Audio concealment

### PR DESCRIPTION
**Pull request by toger5**
## Checklist

-   [x] Tests written for new code (and old code if feasible)
-   [x] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md))

Signed-off-by: Timo Kandra toger5@hotmail.de

---
Here's what your changelog entry will look like:

## ✨ Features
 * Audio concealment ([\#3349](https://github.com/matrix-org/matrix-js-sdk/pull/3349)). Contributed by @toger5.

### CODE REVIEW

File: src/webrtc/stats/summaryStats.ts [hunk 1]
```diff
@@ -24,3 +24,6 @@ export interface TrackSummary {
     maxJitter: number;
     maxPacketLoss: number;
 }
+export interface AudioTrackSummary extends TrackSummary {
```
Comment (EnricoSchw) [inline src/webrtc/stats/summaryStats.ts:27]: I would suggest specialization is not needed. This specialization forces you to do an extra loop:
https://github.com/matrix-org/matrix-js-sdk/blob/fb3bdc39ecfdb12c0d6c3a3dd6e346d771517fde/src/webrtc/stats/trackStatsReporter.ts#L168
And if you not do an extra loop you would need at least an extra  `if - case`


And you initialize a default value anyway. 
https://github.com/matrix-org/matrix-js-sdk/blob/fb3bdc39ecfdb12c0d6c3a3dd6e346d771517fde/src/webrtc/stats/trackStatsReporter.ts#L143
For video this value could be simple 0 or -1

File: src/webrtc/stats/trackStatsReporter.ts [hunk 2]
```diff
@@ -137,26 +137,37 @@ export class TrackStatsReporter {
...
+                trackSummary.maxJitter = stats.getJitter();
+            }
+            if (trackSummary.maxPacketLoss < stats.getLoss().packetsLost) {
+                trackSummary.maxPacketLoss = stats.getLoss().packetsLost;
+            }
+        });
+
+        const audioTrackList = remoteTrackList.filter((t) => t.kind === "audio");
+        audioTrackList.forEach((stats) => {
+            audioTrackSummary.concealedAudioRatio += stats.getAudioConcealment()?.ratio / audioTrackList.length;
```
Comment (EnricoSchw) [inline src/webrtc/stats/trackStatsReporter.ts:169]: I would do this calculation in: https://github.com/matrix-org/matrix-js-sdk/blob/fb3bdc39ecfdb12c0d6c3a3dd6e346d771517fde/src/webrtc/stats/summaryStatsReporter.ts#LL109C8-L109C8

like:

```typescript
summaryCounter.concealedAudio += (stats.audioTrackSummary.count > 0 ? stats.audioTrackSummary.concealedAudioRatio/stats.audioTrackSummary.count : 0)
```

Here i only wold sum the `stats.getAudioConcealment()?.ratio` 

Maybe with my first comment makes also more sense.

**Review by SimonBrandner — COMMENTED** [review]
Looks sane

File: src/rust-crypto/rust-crypto.ts [hunk 3]
```diff
@@ -331,7 +331,7 @@ export class RustCrypto implements CryptoBackend {
             events ? JSON.stringify(events) : "[]",
             devices,
             oneTimeKeysCounts,
-            unusedFallbackKeys,
+            unusedFallbackKeys ?? new Set<string>(),
```
[the remark this answers is not in the document]
Comment (toger5) [reply src/rust-crypto/rust-crypto.ts:334]: yes, i needed to test it to make the test passing...

File: src/webrtc/stats/summaryStats.ts [hunk 4]
```diff
@@ -14,7 +14,7 @@ export interface SummaryStats {
     receivedMedia: number;
     receivedAudioMedia: number;
     receivedVideoMedia: number;
-    audioTrackSummary: TrackSummary;
+    audioTrackSummary: AudioTrackSummary;
     videoTrackSummary: TrackSummary;
```
[the remark this answers is not in the document]
Comment (toger5) [reply src/webrtc/stats/summaryStats.ts]: As enrico proposed I went the other route (only having trackSummary witha ll the props for audio and video)

File: src/webrtc/stats/media/mediaTrackStats.ts [hunk 5]
```diff
@@ -32,7 +33,14 @@ export interface Bitrate {
      */
     upload: number;
 }
+export interface ConcealedAudio {
+    /**
+     * durtation in ms
```
[the remark this answers is not in the document]
Comment (toger5) [reply src/webrtc/stats/media/mediaTrackStats.ts]: thanks!

File: src/webrtc/stats/media/mediaTrackStats.ts [hunk 6]
```diff
@@ -44,6 +52,7 @@ export class MediaTrackStats {
     private loss: PacketLoss = { packetsTotal: 0, packetsLost: 0, isDownloadStream: false };
     private bitrate: Bitrate = { download: 0, upload: 0 };
     private resolution: Resolution = { width: -1, height: -1 };
+    private audioConcealment: AudioConcealment = { ratio: 0, concealedAudio: 0, totalAudioDuration: 0 };
```
[the remark this answers is not in the document]
Comment (toger5) [reply src/webrtc/stats/media/mediaTrackStats.ts:55]: do you mean splitting connectionStats and MediaStats?

**Pull request merged**